ctest lint fixes

maage opened the pull request:
#### Description:

When testing with -DANSIBLE_PLAYBOOKS_PER_RULE_ENABLED=ON -DANSIBLE_CHECKS=ON there was regression. This patchset fixes some of regressions and then comments out rest I decided are not worth not fix now.

#### Rationale:

This allows me to run full ctest run and all tests pass for all products.

This should allow to keep reasonable and similar style within all ansible files. And then later it allows to do bigger refactorings with fqcn and jinja spacing for example. I selected most ansible-lint rules to skip based on issue numbers, so only low number cases were fixed. `name[template]` is added just because I dont't like the rule especially with this kind of sprawling code base with templates and all kinds of cases.

Long lines are not fully fixed because some are ssg template comments and not fixable as is. And also some other points would require not simple fixes. I tried to balance this and then added rule as skipped.

#### Review Hints:

Somehow different pyyaml versions have different serialization formats. So I implemented generic EOL tidy and then kludge to select implementation based on how it output problematic output. This is only tested on Fedora 37.

jan-cerny commented:
> Seems to be unrelated:

I have rerun it. In general, it often is an infrastructure problem. It can be related in some PRs where the PR contain changes that would mean a significant growth or performance hit.

jan-cerny commented on cmake/SSGCommon.cmake:
```diff
@@ -1242,19 +1242,19 @@ macro(ssg_define_guide_and_table_tests)
-        set_tests_properties("unique-cces" PROPERTIES LABELS quick)
-        # set_tests_properties("unique-stigids" PROPERTIES LABELS quick)
     endif()
+    add_test(
+        NAME "unique-cces"
+        COMMAND "${CMAKE_CURRENT_SOURCE_DIR}/tests/assert_reference_unique.py" "${CMAKE_SOURCE_DIR}" "cce"
+    )
```
I seems that the test now doesn't work. I have tried to make some duplicate CCEs and run the test. It prints the results but the result is pass even if they should fail.

```
[jcerny@thinkpad build{pr/10348}]$ ctest --verbose -R unique-cces
UpdateCTestConfiguration  from :/home/jcerny/work/git/scap-security-guide/build/DartConfiguration.tcl
UpdateCTestConfiguration  from :/home/jcerny/work/git/scap-security-guide/build/DartConfiguration.tcl
Test project /home/jcerny/work/git/scap-security-guide/build
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 4
    Start 4: unique-cces

4: Test command: /home/jcerny/work/git/scap-security-guide/tests/assert_reference_unique.py "/home/jcerny/work/git/scap-security-guide" "cce"
4: Working Directory: /home/jcerny/work/git/scap-security-guide/build
4: Test timeout computed to be: 10000000
4: 
4: cce CCE-84079-3 is included in files: 
4:  - linux_os/guide/system/selinux/selinux_policytype/rule.yml
4:  - linux_os/guide/system/selinux/selinux_state/rule.yml
1/1 Test #4: unique-cces ......................   Passed    0.24 sec

The following tests passed:
	unique-cces

100% tests passed, 0 tests failed out of 1

Label Time Summary:
quick    =   0.24 sec*proc (1 test)

Total Test time (real) =   0.25 sec
```

maage replied:
Result code was not implemented. I'll add it and also split main into functions so codeclimate is happier.

maage replied:
Okay, now it passes this edit,
```
1/6 Test  #77: unique-cces.......................***Failed    1.05 sec

cce CCE-84020-7 is included in files: 
 - applications/openshift/api-server/api_server_audit_log_path/rule.yml
 - applications/openshift/api-server/api_server_auth_mode_no_aa/rule.yml
 - applications/openshift/api-server/api_server_auth_mode_node/rule.yml

6/6 Test #283: missing-cces-rhel8
83% tests passed, 1 tests failed out of 6

Label Time Summary:
quick    =  11.08 sec*proc (6 tests)

Total Test time (real) =   2.36 sec

The following tests FAILED:
	 77 - unique-cces (Failed)
Errors while running CTest
```

jan-cerny commented on ssg/yaml.py:
```diff
@@ -21,7 +21,17 @@
+            return super(yaml_test_Dumper, self).increase_indent(flow, False)
+    unformatted_yaml = yaml.dump({"a": [1]}, None, TestDumper)
+    if unformatted_yaml == 'a:\n- 1\n':
+        from yaml import Dumper as yaml_Dumper
```
This will have performance impact - we should prefer the CDumper.

maage replied:
Ok, I'll drop that patch.

maage commented:
Seems `enable_authselect` `remediable.fail.sh` breaks automatus podman test env and maybe others, so I needed to modify it.

Also it seems `dir_perms_world_writable_root_owned` test `world_writable_dir_on_nonlocal_fs.fail.sh` fails because containerized environment might not have nfsd or maybe has other issues too. But I'm not happy with resolution so I dropped that patch. Also dropped `audit_rules_privileged_commands` because I was getting remediation errors and proper fix was too involved.

Added two patches about `test-generate_contributors` and `ssg_controlrefcheck_test` to make ctest more smooth and work as I'd expect.

maage commented:
`unique-cces` test failed under Centos 7, because it used py2
```
-- Found PythonInterp: /usr/bin/python (found version "2.7.5") 
```
added cmake if to guard against py2.

```

maage commented:
Forgot `tests/unit/ssg_test_suite/test_assert_reference_unique.py`, just skip it under py2.
```
# PYTHONPATH=. pytest -v tests/unit/ssg_test_suite/test_assert_reference_unique.py
========================================================================== test session starts ==========================================================================
platform linux2 -- Python 2.7.18, pytest-4.6.11, py-1.11.0, pluggy-0.13.1 -- /usr/bin/python2
cachedir: .pytest_cache
rootdir: /ssg
collected 0 items / 1 skipped                                                                                                                                           

======================================================================= 1 skipped in 0.05 seconds =======================================================================
```

maage commented:
Okay, pytest version in centos-7 env is older than 4.6.x from pip. It seems v2.7.0.
```
platform linux2 -- Python 2.7.5 -- py-1.4.32 -- pytest-2.7.0
```

```
# PYTHONPATH=tests:. python -m pytest -v tests/unit/ssg_test_suite
============================================================= test session starts ==============================================================
platform linux2 -- Python 2.7.5 -- py-1.4.32 -- pytest-2.7.0 -- /usr/bin/python
rootdir: /ssg/tests/unit/ssg_test_suite, inifile: 
collected 27 items / 1 skipped 
```

```
# PYTHONPATH=tests:. python -m pytest -v tests/unit/ssg_test_suite
============================================================= test session starts ==============================================================
platform linux2 -- Python 2.7.5, pytest-4.6.11, py-1.11.0, pluggy-0.13.1 -- /usr/bin/python
cachedir: .pytest_cache
rootdir: /ssg
collected 27 items / 1 skipped / 26 selected                                                                                                   
```

```
# PYTHONPATH=tests:. python3 -m pytest -v tests/unit/ssg_test_suite
============================================================= test session starts ==============================================================
platform linux -- Python 3.6.8, pytest-7.0.1, pluggy-1.0.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /ssg
collected 32 items                                                                                                                             

...
======================================================= 32 passed, 34 warnings in 8.12s ========================================================
```

maage commented:
Looking automatus results.

One way to speed up testing timewise would be to split remediation type per job. For example 3 way works for most: not defined, ansible and bash.

Why "Automatus Fedora / Run Tests (pull_request)" uses "product rhel8" and "ssg-rhel8-ds.xml"? I have used always fedora versions and I guess this explains some discrepancies. Fedora does not have `sssd_ldap_configure_tls_ca_dir` and `service_pcscd_enabled` passes in my test env.

I guess these should be labeled as warnings. They should not fail the test. You are just trying to test a rule on platform not configured to allow rule. Same as **notapplicable**.
```
ERROR - Rule 'xccdf_org.ssgproject.content_rule_ntpd_configure_restrictions' isn't present in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-8' in '/tmp/ssgts-ds-675d98n5'
ERROR - Rule 'xccdf_org.ssgproject.content_rule_accounts_passwords_pam_tally2_unlock_time' isn't present in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-8' in '/tmp/ssgts-ds-675d98n5'
ERROR - Rule 'xccdf_org.ssgproject.content_rule_accounts_passwords_pam_tally2_deny_root' isn't present in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-8' in '/tmp/ssgts-ds-675d98n5'
ERROR - Rule 'xccdf_org.ssgproject.content_rule_accounts_passwords_pam_tally2' isn't present in benchmark 'xccdf_org.ssgproject.content_benchmark_RHEL-8' in '/tmp/ssgts-ds-675d98n5'
```

I think both of these might be unrelated failures. Further analysis would require automatus logs from `logs_ansible` or `logs_bash`.

```
INFO - Script ldap_tls_cacertdir_bad_value.fail.sh using profile xccdf_org.ssgproject.content_profile_stig OK
ERROR - Rule evaluation resulted in fail, instead of expected pass during final stage 
ERROR - The check after remediation failed for rule 'xccdf_org.ssgproject.content_rule_sssd_ldap_configure_tls_ca_dir'.
```

```
ERROR - Script service_enabled.pass.sh using profile (all) found issue:
ERROR - Rule evaluation resulted in error, instead of expected pass during initial stage 
ERROR - The initial scan failed for rule 'xccdf_org.ssgproject.content_rule_service_pcscd_enabled'.
```

maage commented:
Okay, find out why. Because rule is not selected:
```
% oscap --verbose DEVEL xccdf generate fix --result-id xccdf_org.open-scap_testresult_xccdf_org.ssgproject.content_profile_stig --fix-type ansible sssd_ldap_configure_tls_ca_dir-ldap_tls_cacertdir_bad_value.fail.sh-final-arf.xml
I: oscap: Identified document type: asset-report-collection [oscap(309910):oscap(7ce820c25040):doc_type.c:96:oscap_determine_document_type_reader]
I: oscap: Identified document type: data-stream-collection [oscap(309910):oscap(7ce820c25040):doc_type.c:96:oscap_determine_document_type_reader]
I: oscap: Created a new XCCDF session from a SCAP Source Datastream '(null)/report-request.xml'. [oscap(309910):oscap(7ce820c25040):xccdf_session.c:179:xccdf_session_new_from_source]
D: oscap: Validating SCAP Source Datastream (1.3) document from (null)/report-request.xml. [oscap(309910):oscap(7ce820c25040):oscap_source.c:350:oscap_source_validate]
I: oscap: Validating XML signature. [oscap(309910):oscap(7ce820c25040):signature.c:122:_oscap_signature_validate_doc]
I: oscap: Signature node not found [oscap(309910):oscap(7ce820c25040):signature.c:167:_oscap_signature_validate_doc]
D: oscap: SDS session created temporary directory '/tmp/oscap.Y0EXXu'. [oscap(309910):oscap(7ce820c25040):ds_sds_session.c:125:ds_sds_session_get_temp_dir]
D: oscap: Successfully dumped script component (id='scap_org.open-scap_ecomp_rhel8-checks-sce-ssh_keys_passphrase_protected.sh') to file '/tmp/oscap.Y0EXXu/./././rhel8/checks/sce/ssh_keys_passphrase_protected.sh'. [oscap(309910):oscap(7ce820c25040):sds.c:196:ds_sds_dump_component_sce]
WARNING: Datastream component 'scap_org.open-scap_cref_security-data-oval-com.redhat.rhsa-RHEL8.xml.bz2' points out to the remote 'https://access.redhat.com/security/data/oval/com.redhat.rhsa-RHEL8.xml.bz2'. Use '--fetch-remote-resources' option to download it.
WARNING: Skipping 'https://access.redhat.com/security/data/oval/com.redhat.rhsa-RHEL8.xml.bz2' file which is referenced from datastream
I: oscap: Identified document type: Benchmark [oscap(309910):oscap(7ce820c25040):doc_type.c:96:oscap_determine_document_type_reader]
I: oscap: Generating result-oriented fixes for policy(result/@id=xccdf_org.open-scap_testresult_xccdf_org.ssgproject.content_profile_stig) [oscap(309910):oscap(7ce820c25040):xccdf_policy_remediate.c:1301:xccdf_policy_generate_fix]
---
###############################################################################
#
# Ansible Playbook generated from evaluation of DISA STIG for Red Hat Enterprise Linux 8
#
# Profile ID: xccdf_org.ssgproject.content_profile_stig
# XCCDF Version:  1.2
#
# Evaluation Start Time:  2023-03-27T08:17:16+00:00
# Evaluation End Time:  2023-03-27T08:17:16+00:00
#
# This file was generated by OpenSCAP 1.3.8 using:
# $ oscap xccdf generate fix --result-id xccdf_org.open-scap_testresult_xccdf_org.ssgproject.content_profile_stig --fix-type ansible xccdf-results.xml
#
# This Ansible Playbook is generated from the results of a profile evaluation.
# It attempts to remediate all issues from the selected rules that failed the test.
#
# How to apply this Ansible Playbook:
# $ ansible-playbook -i "localhost," -c local playbook.yml
# $ ansible-playbook -i "192.168.1.155," playbook.yml
# $ ansible-playbook -i inventory.ini playbook.yml
#
###############################################################################


- hosts: all
I: oscap: Skipping unselected Rule/@id="xccdf_org.ssgproject.content_rule_sssd_ldap_configure_tls_ca_dir" [oscap(309910):oscap(7ce820c25040):xccdf_policy_remediate.c:819:_xccdf_policy_rule_get_fix_text]
  vars:
  tasks:
```

If I edit arf:
```
% diff -u sssd_ldap_configure_tls_ca_dir-domain_not_there.fail.sh-final-arf.xml xccdf-results.xml
--- sssd_ldap_configure_tls_ca_dir-domain_not_there.fail.sh-final-arf.xml	2023-03-27 11:18:02.678442152 +0300
+++ xccdf-results.xml	2023-03-27 12:55:27.663350563 +0300
@@ -270967,7 +270967,7 @@
                 <xccdf-1.2:check-content-ref href="ssg-rhel8-ocil.xml" name="ocil:ssg-sssd_ldap_configure_tls_ca_ocil:questionnaire:1"/>
               </xccdf-1.2:check>
             </xccdf-1.2:Rule>
-            <xccdf-1.2:Rule selected="false" id="xccdf_org.ssgproject.content_rule_sssd_ldap_configure_tls_ca_dir" severity="medium">
+            <xccdf-1.2:Rule selected="true" id="xccdf_org.ssgproject.content_rule_sssd_ldap_configure_tls_ca_dir" severity="medium">
               <xccdf-1.2:title>Configure SSSD LDAP Backend Client CA Certificate Location</xccdf-1.2:title>
               <xccdf-1.2:description>Configure SSSD to implement cryptography to protect the
 integrity of LDAP remote access sessions. By setting
```
output is expected.

I guess for arf.xml should be modified to mark tested rule as selected. Or maybe actual ds modified to mark rules to be tested as selected. I think any of these ideas should be handled as separate task.

jan-cerny commented:
> Okay, find out why. Because rule is not selected:

The rule not being selected is expected here. But I think that we have hit some bug in OpenSCAP. I have reported it in https://github.com/OpenSCAP/openscap/issues/1963.

jan-cerny commented:
> Okay, find out why. Because rule is not selected:

The rule not being selected is expected here. But I think that we have hit some bug in OpenSCAP. I have reported it in https://github.com/OpenSCAP/openscap/issues/1963.

jan-cerny commented:
> Okay, find out why. Because rule is not selected:

The rule not being selected is expected here. But I think that we have hit some bug in OpenSCAP. I have reported it in https://github.com/OpenSCAP/openscap/issues/1963.
